feat: Add command for config validation (#3220)
grafana/alloy: docs/sources/reference/cli/validate.md

---
canonical: https://grafana.com/docs/alloy/latest/reference/cli/validate/
description: Learn about the validate command
menuTitle: validate
title: The validate command
weight: 200
---

# The `validate` command

The `validate` command validates {{< param "PRODUCT_NAME" >}} configuration file/directory path.

> ## Usage

Reviewer: Can you describe how to use it in CI programmatically, e.g. the non-zero exit code = error, zero exit code = success.

Author: [2a1128d](https://github.com/grafana/alloy/pull/3220/commits/2a1128d1ffc9cb3b6494ca8c01acc9fb1a4dbcd0)

```shell
alloy validate [<FLAG> ...] <PATH_NAME>
```

Replace the following:

* _`<FLAG>`_: One or more flags that define the input and output of the command.
* _`<PATH_NAME>`_: Required. The {{< param "PRODUCT_NAME" >}} configuration file/directory path.

If you give the _`<PATH_NAME>`_ argument a directory path, {{< param "PRODUCT_NAME" >}} finds `*.alloy` files (ignoring nested directories) and loads them as a single configuration source.

The following flags are supported:

* `--config.format`: The format of the source file. Supported formats: `alloy`, `otelcol`, `prometheus`, `promtail`, `static` (default `"alloy"`).
* `--config.bypass-conversion-errors`: Enable bypassing errors when converting (default `false`).
* `--config.extra-args`: Extra arguments from the original format used by the converter.

> * `--stability.level`: The minimum permitted stability level of functionality. Supported values: `experimental`, `public-preview`, `generally-available` (default `"generally-available"`).

Reviewer: Maybe should say that this should be set to the same value you intend to run Alloy with?

Author: Is [7979237](https://github.com/grafana/alloy/pull/3220/commits/79792375b1b2f85e8ad2710be0c5f8d99e1e598c) good enough? Or do you prefer to have this documented on the flag themself?

> * `--feature.community-components.enabled`: Enable community components (default `false`).

Reviewer: & similarly to above

## Limitations

> Validation performed is limited in what it can check. Currently it will check:
> 
> * Syntax errors.
> * Components exist.
> * Components are **unique** across all {{< param "PRODUCT_NAME" >}} configuration file and are not repeated.

Reviewer: ```suggestion
Validation performed is limited in what it can check. Currently it will check for:

* Syntax errors.
* Missing components.
* Components name conflicts.
```

---
canonical: https://grafana.com/docs/alloy/latest/reference/cli/validate/
description: Learn about the validate command
menuTitle: validate
title: The validate command
weight: 200
---

# The `validate` command

> The `validate` command validates {{< param "PRODUCT_NAME" >}} configuration file/directory path.

Author: This was phrased in the same way we have for [run](https://github.com/grafana/alloy/blob/main/docs/sources/reference/cli/run.md?plain=1#L22). If its is a path to a file we only load that file. But if its a directory we try to load all alloy files in that directory

## Usage

```shell
alloy validate [<FLAG> ...] <PATH_NAME>
```

> This command will fail with a non-zero exit code and print diagnostics to stderr if validation fails.

Reviewer 2: Yeah I struggled with this one - what does it do if the configuration is OK. We could tweak it even more here... focus furst on a success condition, then explain what happens if it fails validation.

```suggestion
If the configuration file is valid, the `validate` command returns a zero exit code.
If the configuration file is invalid, the command  returns a non-zero exit code and prints diagnostics generated during validation to stderr.
```

Replace the following:

* _`<FLAG>`_: One or more flags that define the input and output of the command.
* _`<PATH_NAME>`_: Required. The {{< param "PRODUCT_NAME" >}} configuration file/directory path.

> If you give the _`<PATH_NAME>`_ argument a directory path, {{< param "PRODUCT_NAME" >}} finds `*.alloy` files (ignoring nested directories) and loads them as a single configuration source.

Author: This was taken from https://github.com/grafana/alloy/blob/main/docs/sources/reference/cli/run.md?plain=1#L26. Should we update there too?

> If you give the _`<PATH_NAME>`_ argument a directory path, {{< param "PRODUCT_NAME" >}} finds `*.alloy` files (ignoring nested directories) and loads them as a single configuration source.

Author: Did it here [224c91d](https://github.com/grafana/alloy/pull/3220/commits/224c91d8c6d9eadd195b16be6d2ff8bd8bda0d1e)

The following flags are supported:

> * `--config.format`: The format of the source file. Supported formats: `alloy`, `otelcol`, `prometheus`, `promtail`, `static` (default `"alloy"`).

Reviewer 3: Maybe we should have some kind of clear message added to the log in this case, otherwise I think this could be very confusing should a user ever attempt it.

> * `--config.format`: The format of the source file. Supported formats: `alloy`, `otelcol`, `prometheus`, `promtail`, `static` (default `"alloy"`).
> * `--config.bypass-conversion-errors`: Enable bypassing errors when converting (default `false`).
> * `--config.extra-args`: Extra arguments from the original format used by the converter.
> * `--stability.level`: The minimum permitted stability level of functionality. Supported values: `experimental`, `public-preview`, `generally-available` (default `"generally-available"`).

Reviewer 2: ```suggestion
* `--config.format`: Specifies the source file format. Supported formats: `alloy`, `otelcol`, `prometheus`, `promtail`, and `static` (default `"alloy"`).
* `--config.bypass-conversion-errors`: Enable bypassing errors during conversion (default `false`).
* `--config.extra-args`: Extra arguments from the original format used by the converter.
* `--stability.level`: The minimum permitted stability level of functionality. Supported values: `experimental`, `public-preview`, and `generally-available` (default `"generally-available"`).
```

Author: [d036be6](https://github.com/grafana/alloy/pull/3220/commits/d036be64b91cad728f0376b37bd68e0e828ea9a4)

* `--feature.community-components.enabled`: Enable community components (default `false`).

{{< admonition type="note" >}}

> `--stability.level` and `--feature.community-components.enabled` should be set to the same value that is used when running alloy.

Reviewer 2: ```suggestion
Set `--stability.level` and `--feature.community-components.enabled` to the same value used when running {{< param "PRODUCT_NAME" >}}.
```

> `--stability.level` and `--feature.community-components.enabled` should be set too the same values that is used when running alloy.

Reviewer: ```suggestion
`--stability.level` and `--feature.community-components.enabled` should be set to the same value that is used when running alloy.
```

{{< /admonition >}}

## Limitations

> Validation performed is limited in what it can check. Currently it will check for:

Reviewer 2: ```suggestion
Validation is limited in scope. It currently checks for:
```

* Syntax errors.
* Missing components.

> * Components name conflicts.

Reviewer 2: ```suggestion
* Component name conflicts.
```

Replace the following:

* _`<FLAG>`_: One or more flags that define the input and output of the command.
* _`<PATH_NAME>`_: Required. The {{< param "PRODUCT_NAME" >}} configuration file or directory path.

If you provide a directory path for  the _`<PATH_NAME>`_, {{< param "PRODUCT_NAME" >}} finds `*.alloy` files, ignoring nested directories, and loads them as a single configuration source.

The following flags are supported:

* `--config.format`: Specifies the source file format. Supported formats: `alloy`, `otelcol`, `prometheus`, `promtail`, and `static` (default `"alloy"`).
* `--config.bypass-conversion-errors`: Enable bypassing errors during conversion (default `false`).
* `--config.extra-args`: Extra arguments from the original format used by the converter.
* `--stability.level`: The minimum permitted stability level of functionality. Supported values: `experimental`, `public-preview`, and `generally-available` (default `"generally-available"`).
* `--feature.community-components.enabled`: Enable community components (default `false`).

{{< admonition type="note" >}}

> Set `--stability.level` and `--feature.community-components.enabled` to the same value used when running {{< param "PRODUCT_NAME" >}}.

Reviewer 2: ```suggestion
When you validate the {{< param "PRODUCT_NAME" >}} configuration, you must set the `--stability.level` and `--feature.community-components.enabled` arguments to the same values you want to use when you run {{< param "PRODUCT_NAME" >}}.
```

This one is a bit confusing on first read. Maybe something like this small change will make it more clear?

Author: Yes I agree, this is better

> {{< /admonition >}}
> 
> 
> ## Limitations

Reviewer 2: ```suggestion
{{< /admonition >}}

## Limitations
```

Remove extra empty line
